Switch to first-class aggregate support in pgx

JamesGuthrie opened the pull request:
Note this is a WIP, and removes backwards-compatibility functions.

Part of https://github.com/timescale/promscale_extension/issues/61

JamesGuthrie committed:
Switch to first-class aggregate support in pgx

- Pgx automatically handles in_aggregate_context, so we can remove some
  (now) unused code.
- Because vector_selector's state type is no longer INTERNAL, we don't
  need to provide (de)serialization functions.
- We have removed backwards-compatibility wrappers for aggregates.

bitwalker commented on src/aggregates/prom_increase.rs:
```diff
@@ -1,109 +1,50 @@
-
-            Some(state)
-        })
+pub struct PromIncrease {}
```
NOTE: You can declare zero-sized structs using just `pub struct PromIncrease;`

bitwalker commented on src/aggregates/prom_increase.rs:
```diff
@@ -1,109 +1,50 @@
+#[pg_aggregate]
+impl Aggregate for PromIncrease {
+    type State = Option<GapfillDeltaTransition>;
+    type Args = (
```
It's unfortunate that this trait requires an extra allocation just to pack the arguments into a tuple in order to call the `state` function, but I guess that was a tradeoff they were forced to make. I wonder what the impact of that is in a hot aggregate.

bitwalker reviewed (commented):
Definitely cleaner! I think we'll want to try and benchmark the overhead of wrapping arguments in a tuple, since I could see that potentially being an issue in hot aggregates, but we'll have to concoct a decent benchmark setup for that (we probably should be thinking about that anyway, especially with performance being one of our main objectives this year). 

Rust does have facilities for that, but we may have to upstream some support in pgx for it, since it would need to build on top of the pgx infra, just like how testing does.

JamesGuthrie committed:
Bump pgx to 0.3.1

JamesGuthrie committed:
Fix broken build on develop branch

`#[pgx(sql = false)]` suppresses SQL generation for this object, so the
generated SQL is incomplete. We will use this SQL-suppression, but it's
missing some pieces which still need to be developed.
